src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1272,39 +1276,46 @@ private ArrayTypeSymbol VisitArrayInitializer(BoundArrayCreation node)
-                // Should do the same here.
-                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                // If there are error types, use the first error type. (Matches InferBestType(ImmutableArray<BoundExpression>, ...).)
-                var bestType = resultTypes.FirstOrDefault(t => !t.IsNull && t.IsErrorType());
-                if (bestType.IsNull)
+                if (node.HasErrors)
                 {
-                    bestType = BestTypeInferrer.InferBestType(resultTypes, _conversions, useSiteDiagnostics: ref useSiteDiagnostics);
+                    elementType = elementType.SetUnknownNullabilityForReferenceTypes();
+                    checkConversions = false;
                 }
                 else
                 {
-                    // Mark the error type as null-oblivious.
-                    bestType = TypeSymbolWithAnnotations.Create(bestType.TypeSymbol);
-                }
-                // https://github.com/dotnet/roslyn/issues/27961 Report a special ErrorCode.WRN_NoBestNullabilityArrayElements
-                // when InferBestType fails, and avoid reporting conversion warnings for each element in those cases.
-                // (See similar code for conditional expressions: ErrorCode.WRN_NoBestNullabilityConditionalExpression.)
-                if (!bestType.IsNull)
-                {
-                    elementType = bestType;
+                    var placeholderBuilder = ArrayBuilder<BoundExpression>.GetInstance(n);
+                    for (int i = 0; i < n; i++)
+                    {
+                        placeholderBuilder.Add(CreatePlaceholderIfNecessary(elementBuilder[i], resultBuilder[i]));
+                    }
+                    var placeholders = placeholderBuilder.ToImmutableAndFree();
+                    HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                    var bestType = BestTypeInferrer.InferBestType(placeholders, _conversions, out _, ref useSiteDiagnostics);
+                    if (bestType is null)
+                    {
+                        ReportDiagnostic(ErrorCode.WRN_NoBestNullabilityArrayElements, node.Syntax);
+                        bestType = elementType.TypeSymbol.SetUnknownNullabilityForReferenceTypes();
+                        checkConversions = false;
+                    }
+                    else
+                    {
+                        checkConversions = true;
+                    }
+                    elementType = TypeSymbolWithAnnotations.Create(bestType, isNullableIfReferenceType: BestTypeInferrer.GetIsNullable(resultBuilder));
                 }
                 arrayType = arrayType.WithElementType(elementType);
             }
+            else
+            {
+                checkConversions = true;
+            }
 
-            if (!elementType.IsNull && !elementType.IsValueType)
+            if (checkConversions && !elementType.IsNull && !elementType.IsValueType)
```

Comment:
>!elementType.IsNull && [](start = 36, length = 22)

I'm not sure if we still need to check for default `elementType`. `checkConversions` seems sufficient.